### PR TITLE
transfer loading Modal

### LINKED ISSUE #6467: Transfer Redesign handover - (3) Transaction loading modal
related to #6460

Task:
- [ ] desktop redesign of these part
image

figma: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?node-id=890%3A14850&mode=dev

### LINKED ISSUE #6944: netwrok 'reset' after transfer
### What happened?

After successful transfer of funds
chosen Token / network is reset back to KSM
it's annoying

### Please reproduce in steps

go to transfer page
change Token to DOT
transfer away
Form is reset back to KSM

### Expected Behavior

Token stays same

### What browsers are you seeing the problem on?

Desktop Firefox

### At which address did you encounter bug?

_No response_

### Are you logged in?

Yes

### Which wallet you are using?

Talisman

### At which chain did you encounter bug?

_No response_

### Screenshots

no

### Relevant log output

_No response_

### Payment link for reward

https://kodadot.xyz/ksm/transfer/?target=%3CMy_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address%3E

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by daiagi**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [x] Feature
- [x] Refactoring

## Needs Design check

- @exezbcz please review

## Needs QA check

- @kodadot/qa-guild please review

## Context

- [x] Closes #6467
- [x] Closes #6944 
- [x] refactor: replace `status` returned from `useTransactinStatus` by enum instead of strings
        side effect change in `Buy.vue` and `Massmint.vue` 
- [x] bug fix:  console warnings from `AddressInput` about required `emptyOnError` prop
- [x] bug fix:  console warnings from `AddressInput` about undefined v-model value (Transfer page, change `targetAddresses.address` empty state value from `undefined` to empty string `''`

#### Did your issue had any of the "$" label on it?

Yes. **$$**

- [x] My DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)

## Screenshot 📸

- [x] My fix has changed UI

image

image

image

### <samp>🤖 Generated by Copilot at d0be5a5</samp>

This pull request introduces a new `TransactionLoader` component that shows the status and result of transactions in a modal, and updates the `Buy`, `Transfer` and `Massmint` components to use it. It also improves the code organization, type safety, user interface and customization of several components and composables, such as `AddressInput`, `NeoModal`, `NeoSteps`, `useChain` and `useTransactionStatus`. It adds new translation keys and values for the `TransactionLoader` component to `en.json`.

### <samp>🤖 Generated by Copilot at d0be5a5</samp>

> _Sing, O Muse, of the glorious deeds of the code reviewers,_
> _Who with skill and wisdom improved the transaction process,_
> _Using `TransactionStatus` enum and `TransactionLoader` modal,_
> _And enhancing the user interface with `NeoSteps` and `NeoModal`._

### CODE REVIEW

**Comment by daiagi** [comment]
@exezbcz 
few things from the design did not make it into this PR:

1.  cancel button:
image


if I understood you correctly. this is intended to cancel the transaction, as can be done via the wallet sign modal
This, AFAIK, is not possible.
please @kodadot/internal-dev  correct me if I'm wrong

2. Transaction time estimate:
image


This too, I don't know how and if it can be done

**Comment by daiagi** [comment]
what am I supposed to do about this?

image

**Comment by Jarsen136** [comment]
> what am I supposed to do about this?
> 
> image

`v-safe-href` could help

**Comment by Jarsen136** [comment]
> 1. cancel button:
>    image
> 
> if I understood you correctly. this is intended to cancel the transaction, as can be done via the wallet sign modal This, AFAIK, is not possible.

I thought the button was used to close the loading modal and also the wallet sign modal. Users could click the button before signing tx and sending the tx on-chain.

**Comment by daiagi** [comment]
> `v-safe-href` could help

image


tried it on local machine
The button stopped working (i.e. not linking anywhere)


is it really a security issue?
it is not a user-provided URL

**Comment by vikiival** [comment]
> if I understood you correctly. this is intended to cancel the transaction, as can be done via the wallet sign modal

Yup you are correct once you wake up the sign-popup it has to be cancelled by the extension

> transaction time

Luckily i can help you with that!

If you see loader has 3 steps 
1. Broadcast
2. In block 
3. Finalization

Block production takes around 12sec on parachains (bsx, ah*)
6 sec on relaychain (ksm, rmrk, dot)

Som from beginiing you can estimate 3*blocktime and with each step (we call it status) you decrease by blocktime

**Comment by daiagi** [comment]
Thanks viki

Let me sun up the action items

1. Cancel transaction button => can't be done

2. I'll implement time estimated as you suggested. Thanks

**Comment by Jarsen136** [comment]
> > `v-safe-href` could help
> 
> image
> 
> tried it on local machine The button stopped working (i.e. not linking anywhere)
> 

How about `v-safe-href="explorerLink"` ?

It should work as it has been used in other place.
image

**Comment by exezbcz** [comment]
yes! 
that is good, thank you 

> 1. cancel button:

its okay, if its not possible, the modal is functional anyway

image

one thing is dark mode: 

### Dark mode 
- sorry, I did not include that since the signing modals are not finished 100%. Can you please match the styling for darkmode like this? 
image


nice job! thanks

**Comment by daiagi** [comment]
1. Estimated time  =>  Done
2. Dark mode => Done



One note @exezbcz :

in the Figma file, the divider color (the line between 1 step to the next) is:
light mode: shade (#CCCCCC)
dark mode: k-hovergrey (#6B6B6B)

We currently don't have this light/dark pair in our theme file
https://github.com/kodadot/nft-gallery/blob/main/libs/ui/src/scss/_theme.scss

we have this pair:
k-shade:
light: (#CCCCCC)
dark: (#999999)


in an attempt for consistency across the app and to restrict the unique options in the theme file I have chosen to use the existing k-shade, **which means the dark mode color is #999999 and not #6B6B6B** 

I hope that's OK with you, if not let me know and I will add a new color pair to the theme

Thank you :)

**Comment by prury** [comment]
working smoothly ✔️ 
made some transfers ✔️ 
selected asset keeps selected after transferring ✔️

**Comment by exezbcz** [comment]
@daiagi ye its oke, thanks!

File: components/buy/Buy.vue [hunk 1]
```diff
@@ -48,7 +48,10 @@ const ShoppingCartItemToTokenToBuy = (item: ShoppingCartItem): TokenToBuy => {
 }
 
 watchEffect(() => {
-  if (isLoading.value === false && status.value === 'loader.finalized') {
+  if (
+    isLoading.value === false &&
```
Comment (stephenjason89) [inline components/buy/Buy.vue:52]:
```suggestion
    !isLoading.value &&
```

File: composables/useTransactionStatus.ts [hunk 2]
```diff
@@ -1,39 +1,55 @@
 import { ExtrinsicStatus } from '@polkadot/types/interfaces'
 
+export enum TransactionStatus {
+  Broadcast = 'loader.broadcast',
+  Casting = 'loader.casting',
+  Sign = 'loader.sign',
+  Block = 'loader.block',
+  Finalized = 'loader.finalized',
+  Unkonwn = '',
```
Comment (stephenjason89) [inline composables/useTransactionStatus.ts:9]:
```suggestion
  Unknown = '',
```

File: composables/useTransactionStatus.ts [hunk 3]
```diff
@@ -1,39 +1,55 @@
...
+  IPFS = 'loader.ipfs',
+}
+
 function useTransactionStatus() {
-  const status = ref('')
+  const status = ref<TransactionStatus>(TransactionStatus.Unkonwn)
```
Comment (stephenjason89) [inline composables/useTransactionStatus.ts:14]:
```suggestion
  const status = ref<TransactionStatus>(TransactionStatus.Unknown)
```

File: composables/useTransactionStatus.ts [hunk 4]
```diff
@@ -1,39 +1,55 @@
...
     if (extrinsicStatus.isFinalized) {
-      status.value = omitFinalized ? '' : 'loader.finalized'
+      status.value = omitFinalized
+        ? TransactionStatus.Unkonwn
+        : TransactionStatus.Finalized
       return
     }
 
-    status.value = ''
+    status.value = TransactionStatus.Unkonwn
```
Comment (stephenjason89) [inline composables/useTransactionStatus.ts:42]:
```suggestion
    status.value = TransactionStatus.Unknown
```

File: composables/useTransactionStatus.ts [hunk 5]
```diff
@@ -1,39 +1,55 @@
...
   const initTransactionLoader = (): void => {
     isLoading.value = true
-    status.value = 'loader.sign'
+    status.value = TransactionStatus.Unkonwn
   }
 
   const stopLoader = (): void => {
     isLoading.value = false
-    status.value = ''
+    status.value = TransactionStatus.Unkonwn
```
Comment (stephenjason89) [inline composables/useTransactionStatus.ts:52]:
```suggestion
    status.value = TransactionStatus.Unknown
```

**Comment by stephenjason89** [comment]
Code lgtm besides some minor typos

File: composables/useBlockTime.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import { Prefix } from '@kodadot1/static'
+
+export default function () {
+  const { urlPrefix } = usePrefix()
+
+  const paraChainBlockTime = 12 //seconds
+  const relayChainBlockTime = 6 //seconds
+
+  const chainBlockTimes: { [K in Prefix]: number } = {
```
Comment (vikiival) [inline composables/useBlockTime.ts:9]:
```suggestion
  const chainBlockTimes: PartialConfig<number> = {
```

File: composables/useBlockTime.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    ksm: relayChainBlockTime,
+    dot: relayChainBlockTime,
+    rmrk: relayChainBlockTime,
+    ahk: paraChainBlockTime,
+    ahp: paraChainBlockTime,
+    bsx: paraChainBlockTime,
+    snek: paraChainBlockTime,
+    movr: paraChainBlockTime,
+    glmr: paraChainBlockTime,
```
Comment (vikiival) [inline composables/useBlockTime.ts:18]:
```suggestion
    ksm: relayChainBlockTime,
    dot: relayChainBlockTime,
    rmrk: relayChainBlockTime,
```

File: composables/useBlockTime.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  const blocktime = computed(() => chainBlockTimes[urlPrefix.value])
```
Comment (vikiival) [inline composables/useBlockTime.ts:21]:
```suggestion
  const blocktime = computed(() => chainBlockTimes[urlPrefix.value] || paraChainBlockTime)
```

**Review by vikiival — CHANGES_REQUESTED** [review]
Smol other stuff pointed out by @stephenjason89

File: composables/useTransactionStatus.ts (hunk 5 above)
Comment (daiagi) [reply composables/useTransactionStatus.ts:52]: Thank you. Will fix the typo

Comment (daiagi) [reply composables/useTransactionStatus.ts]: done

**Comment by daiagi** [comment]
@vikiival  your review has been implemented. Thanks
@prury  all yours

**Comment by prury** [comment]
loading animations plays correctly ✔️ 
loading steps works as intended ✔️ 
show transactions leads to the correct sub scan ✔️ 
time estimation works nicely ✔️ 
dark and white mode ok ✔️ 

will make follow up for mobile!

**Comment by EthVlad** [comment]
> > 1. cancel button:
> >    image
> > 
> > if I understood you correctly. this is intended to cancel the transaction, as can be done via the wallet sign modal This, AFAIK, is not possible.
> 
> I thought the button was used to close the loading modal and also the wallet sign modal. Users could click the button before signing tx and sending the tx on-chain.

Were you able to implement this or it was not indeed possible?

**Comment by daiagi** [comment]
>Were you able to implement this or it was not indeed possible?

not possible

**Review by stephenjason89 — APPROVED** [review]

**Comment by yangwao** [comment]
pay 150 usd
perfect!

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $150 @ 4.58 USD/DOT ~ 32.751 $DOT
🧗 [1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8](https://polkadot.subscan.io/account/1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)
🔗 [0xf2ddcb63652f602090d00ba77dc978a00b280ffb8f37eb20b548dac3561e3223](https://polkadot.subscan.io/extrinsic/0xf2ddcb63652f602090d00ba77dc978a00b280ffb8f37eb20b548dac3561e3223)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**